### PR TITLE
Split Modal into separate TS files

### PR DESCRIPTION
**Pull request by oisincoveney**
### Description

Splits the Modal component into separate files for each of the separate modal components defined in `Modal.jsx`, and converts each file to typescript.

### How to verify

The Modal should work exactly as before, so no tests should be affected.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30845)

### CODE REVIEW

**Comment by oisincoveney** [comment]
@kulyk here's the TS refactor, will just wait for tests to go through and make any fixes before review

File: frontend/src/metabase/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import { RoutelessFullPageModal } from "metabase/components/Modal/RoutelessFullPageModal";
+import {
+  WindowModal,
+  WindowModalProps,
+} from "metabase/components/Modal/WindowModal";
+import { FullPageModalProps } from "metabase/components/Modal/FullPageModal";
```
Comment (iethree) [inline frontend/src/metabase/components/Modal/Modal.tsx:7]: I'm a big fan of using `import type` wherever possible. It makes reading the imports a lot easier.

File: frontend/src/metabase/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  ) : (
+    <WindowModal {...props} />
+  );
+
+Modal.defaultProps = {
+  isOpen: true,
+};
+
+// eslint-disable-next-line import/no-default-export -- deprecated usage
+export default Modal;
```
Comment (iethree) [inline frontend/src/metabase/components/Modal/Modal.tsx:29]: can we change this to a named export without breaking lots of other imports? we can still export it as default from the index file.

File: frontend/src/metabase/components/Modal/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  full?: boolean;
+  isOpen?: boolean;
+} & (WindowModalProps & FullPageModalProps)) =>
+  full ? (
+    props.isOpen ? (
+      <RoutelessFullPageModal {...props} />
+    ) : null
+  ) : (
+    <WindowModal {...props} />
+  );
```
Comment (iethree) [inline frontend/src/metabase/components/Modal/Modal.tsx:22]: some if statements would probably be a bit easier to read than this nested ternary

**Review by iethree — APPROVED** [review]
🏆  awesome work converting this to typescript. Love to see this kind of work 🏆 

Couple style nits below that it would be nice to address if they make sense to you.

File: frontend/src/metabase/components/Modal/Modal.tsx (hunk 1 above)
Comment (oisincoveney) [reply frontend/src/metabase/components/Modal/Modal.tsx:7]: Ah I didn't know we even used that! I've always been a fan of type imports as well

File: frontend/src/metabase/components/Modal/Modal.tsx (hunk 3 above)
Comment (oisincoveney) [reply frontend/src/metabase/components/Modal/Modal.tsx:22]: Absolutely agree 😄

**Pull request merged**